# **Tutorial : From Searching Models to Simulation and Asking Questions**

This tutorial will walk you through the process of getting started with the useage of different tools to operate with Systems biology models. It includes installation instructions, a quick start guide, and an overview of how to integrate and use the tool in your projects.

## **Getting Started: Quick Start**

Install aiagents4pharma
To begin, you need to install the aiagents4pharma package, which includes the search_models tool for querying biological models.

Use the following command to install the package via pip:

In [165]:
# !pip install aiagents4pharma 
#This will install the aiagents4pharma library and all its dependencies.

**Set Up Your API Key**

Before using the model tools, you need to authenticate using an API key. You can either store the API key in an environment variable or set it directly in your notebook/script.

### Option 1 : Retrieve API Key from Environment Variable

In [2]:
import os
import openai 

# Retrieve API key from environment variable
openai.api_key = os.getenv("OPENAI_API_KEY")

### Option 2: Store API Key Directly in the Script (Temporary)

In [ ]:
import os

os.environ["OPENAI_API_KEY"] = "Your_API_KEY"
# Make sure to replace "your_api_key" with your actual API key.

***Once you have the package installed and the API key set up, you can start using the model tools to query biological models. Here's how you can get started quickly:***
## **Quick Start with search_models Tool**
### 1. Import Required Packages
This imports the SearchModelsTool class from the aiagents4pharma package.

In [ ]:
# Import the `SearchModelsTool` class from search_models.py
from aiagents4pharma.talk2biomodels.tools.search_models import SearchModelsTool

### 2. Initialize the Search Models Tool

This creates an instance of the SearchModelsTool which you will use to search for biological models.

In [ ]:
# Initialize the SearchModelsTool
search_tool = SearchModelsTool()

### 3. Run the Search Models Tool

In this example, we are searching for models related to Crohn's disease. You can replace the query with any disease or biological term you're interested in.

In [272]:
# Define the search query as follows
query = "crohn's disease"

# Run the tool with the query
search_results = search_tool.invoke(input={"query": query})

### 4. Display the Results

To view the results, simply print them

The search results will be displayed in a table format containing the Model ID, Model Name, Format, and Submission Date.

In [273]:
print(search_results)

| #  | BioModel ID                                                                 | BioModel Name                                                                                                                    | Format | Submission Date |
|----|-----------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------|--------|------------------|
| 1  | [BIOMD0000000534](https://www.ebi.ac.uk/biomodels/BIOMD0000000534)        | Dwivedi2014 - Healthy Volunteer IL6 Model                                                                                     | SBML   | 2014-08-04       |
| 2  | [BIOMD0000000535](https://www.ebi.ac.uk/biomodels/BIOMD0000000535)        | Dwivedi2014 - Crohns IL6 Disease model - Anti-IL6 Antibody                                                                   | SBML   | 2014-08-04       |
| 3  | [BIOMD0000000536](https://www.eb

## **Quick Start with Model_Description Tool**

### 1. Import Required Modules
The purpose of this step is to import the necessary classes to work with biological models. The ModelDescriptionTool is used to query and retrieve information about a biological model, and ModelData is a helper class to provide the model's details like its ID or file path. These imports are essential for interacting with the tool.

In [245]:
from aiagents4pharma.talk2biomodels.tools.model_description import ModelDescriptionTool,ModelData

### 2. Initialize the Tool
This step initializes the ModelDescriptionTool, which will allow you to interact with the BioModels database. It sets up the tool to process your questions and fetch the relevant answers from the specified model.

In [246]:
tool = ModelDescriptionTool()

### 3. Define the Model Data
The purpose of defining the model data is to specify the biological model you want to work with. By providing the modelid, you indicate which model from the BioModels database the tool should analyze. Alternatively, you can use the sbml_file_path to load a local SBML file, but here we focus on using a BioModels ID.

In [247]:
model_data = ModelData(
    modelid=537,  # Use a valid BioModels ID
    sbml_file_path=None
)

### 4. Prepare Input, Invoke the Tool, and Display the Result

In this step, you define a question about the biological model and link it to the specified model data. The input is processed by the tool's invoke method, which retrieves the relevant information, and the response is displayed, providing a clear answer to your query

In [248]:
input_data = {
    "question": "What biological processes does this model simulate?",
    "sys_bio_model": model_data
}
response = tool.invoke(input=input_data)
print("Response:", response)


None st_session_key
Response: This model simulates interleukin-6 (IL-6)-mediated immune regulation in Crohn's disease. Specifically, it focuses on the differential activity of various ligands on the signaling of IL-6, which in turn affects the activity of Signal Transducer and Activator of Transcription 3 (STAT3) on the production of the biomarker C-Reactive Protein (CRP) expression. The model integrates intracellular signaling with organ-level dynamics of pharmacological markers to study the impact of different biotherapeutic strategies targeting IL-6-mediated signaling in Crohn's disease.


As you can see, the query's answer is displayed in a clear and well-structured format.

## **Quick Start with simulate_model Tool**
Simulation allows us to run virtual experiments on biological systems. Instead of conducting physical experiments in the lab, we can use computational models to predict how a system behaves under specific conditions.

**Let's simulate a virtual patient with a CRP value of 100 and observe the results over time**

### 1. Import Required Packages

In [279]:
from aiagents4pharma.talk2biomodels.tools.simulate_model import SimulateModelTool,TimeData,SpeciesData,TimeSpeciesNameConcentration
import streamlit as st

SimulateModelTool: The main tool for running model simulations.

TimeData, SpeciesData, TimeSpeciesNameConcentration: Helper classes to define time, species, and recurring data.

streamlit: A Python library for building web applications, used here to manage session data and visualize results.

### 2. Initialize the Tool
Create an instance of the SimulateModelTool.

In [280]:

st.session_state["test_key"] = None

2024-12-10 00:32:53.720 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:32:53.721 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


st.session_state: Stores data during the session. Here, we initialize it with a key (test_key) to save simulation results.

In [281]:
simulate_tool = SimulateModelTool(return_direct=True,st_session_key="test_key")

simulate_tool: Initializes the simulation tool and links it to the Streamlit session using st_session_key.

### 3. Define the Input Data
Define the input data for the simulation in a structured format. You'll provide the model data, time data, species data, and recurring events (if any).

In [282]:
# Define time data
model_data=ModelData(
        modelid= 537,  # Use a valid BioModels ID
        sbml_file_path=None
)
time_data = TimeData(
    duration=100.0,  # Duration of the simulation
    interval=10      # Interval between time steps
)

# Define species data
species_data = SpeciesData(
    species_name=["CRP{serum}"],
    species_concentration=[100.0]
)

# Define recurring data
new_data = [
    TimeSpeciesNameConcentration(time=50, species_name="CRP{serum}", species_concentration=2.0),
]

### 4. Run the Simulation Using invoke
Now, run the simulation using the invoke method. This method will use the structured input data to execute the simulation.

In [283]:
reponse = simulate_tool.invoke(input ={
    "model_data": {"modelid" : model_data.modelid},
    "time_data": {"duration" : time_data.duration,
                    "interval" : time_data.interval},
    "species_data" : {"species_name" : species_data.species_name,
                    "species_concentration" : species_data.species_concentration},
    "recurring_data"  : {"data": new_data}
    })

2024-12-10 00:33:00.487 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:33:00.488 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:33:02.793 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:33:02.795 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


### 5. Display the Results
Finally, display the simulation results in the Jupyter notebook using print.

In [284]:
df =st.session_state["test_key"].simulation_results
# Retrieve and display simulation results from st.session_state

2024-12-10 00:33:03.180 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


In [ ]:
df["CRP[serum]"]
# For example, the  rows of CRP concentration can be shown

0     100.000000
1     210.456232
2     217.986846
3     219.501977
4     219.271887
5     218.450524
6     198.450980
7     211.093574
8     213.313174
9     212.665273
10    211.131418
Name: CRP[serum], dtype: float64

## **Quick Start with AskQuestion Tool** 

### 1. Import Required Packages
Ensure that the required package (aiagents4pharma) is installed and imported correctly.

In [293]:
from aiagents4pharma.talk2biomodels.tools.ask_question import AskQuestionTool, AskQuestionInput

### 2. Initialize the Tool
Create an instance of the AskQuestionTool.

In [294]:
ask_question_tool = AskQuestionTool(return_direct=True,st_session_key="test_key")

### 3. Define the Input Question
Define the input question in the required structure.

In [295]:
# Prepare the input for invoking the tool
input_data = {
    "question": "What is the concentration of CRP[serum] at time 50?",
    "sys_bio_model": "simulation_results"
}

### 4. Run the Question Using invoke
Now, use the invoke method to get the answer to the question.

In [296]:
# Invoke the tool and get the response
response_ = ask_question_tool.invoke( input=input_data)

2024-12-10 00:46:10.123 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:46:10.124 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:46:10.125 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:46:10.126 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:46:10.128 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:46:10.129 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:46:10.139 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2024-12-10 00:46:10.140 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bar

### 5. Display the Answer
Finally, display the answer in the Jupyter notebook.

In [297]:
# Print the response
print(response_)

The concentration of CRP[serum] at time 50 is 218.450524.


In [298]:
df["CRP[serum]"].head(6)

0    100.000000
1    210.456232
2    217.986846
3    219.501977
4    219.271887
5    218.450524
Name: CRP[serum], dtype: float64

At the timestamp 5, the value obtained is 218.450524. Upon verification:

- The LLM's response matches this value accurately.
- The value is consistent with the corresponding entry in the dataframe (df) above.

This alignment confirms that the model is functioning correctly, as the tool's output is validated against the simulation results.